src/main/java/org/aksw/simba/squirrel/deduplication/hashing/impl/IntervalBasedMinHashFunction.java

```diff
@@ -0,0 +1,64 @@
+package org.aksw.simba.squirrel.deduplication.hashing.impl;
+
+import org.aksw.simba.squirrel.deduplication.hashing.HashValue;
+import org.aksw.simba.squirrel.deduplication.hashing.RDFHashFunction;
+import org.apache.jena.graph.Triple;
+
+import java.util.List;
+
+
+/**
+ * Use Min hash algorithm to compute hash values for triples.
+ */
+public class IntervalBasedMinHashFunction implements RDFHashFunction {
+
+    /**
+     * The number n for the 2^n Intervals for the Hashing.
+     */
+    private int powerNumberOfIntervals;
+
+    /**
+     * Constructor.
+     *
+     * @param powerNumberOfIntervals The powered number of Intervalls for the hashing.
+     */
+    public IntervalBasedMinHashFunction(int powerNumberOfIntervals) {
+        this.powerNumberOfIntervals = powerNumberOfIntervals;
+    }
+
+    @Override
+    public HashValue hash(List<Triple> triples) {
+        Integer[] hashValues = new Integer[(int) Math.pow(2, powerNumberOfIntervals)];
+
+        for (Triple triple : triples) {
+            if (triple.getObject().isBlank() || triple.getSubject().isBlank()) {
+                continue;
+            }
+            int hash = triple.hashCode();
+
+            int bitShiftedNumber = hash >>> (32 - powerNumberOfIntervals);
+            String shortBinaryString = Integer.toBinaryString(bitShiftedNumber);
+            StringBuilder binaryString = new StringBuilder();
+
+            //fill with zeros
+            for (int j = 0; j < powerNumberOfIntervals - shortBinaryString.length(); j++) {
+                binaryString.append("0");
+            }
+            binaryString.append(shortBinaryString);
+            String lastBits = binaryString.toString().substring(0, powerNumberOfIntervals);
+
+            //if we have only one interval, the zero is the desired interval
+            if (lastBits.equals("")) {
+                lastBits = "0";
+            }
+
+            int intervalNumber = Integer.parseInt(lastBits, 2);
+
+            if (hashValues[intervalNumber] == null || hashValues[intervalNumber] > hash) {
+                hashValues[intervalNumber] = hash;
+            }
+        }
```

Comment:
integer -> string -> integer :scream: 

That does not look correct and the bit shifting is not working as it should. Please make sure that this gets fixed. Additionally, I would like to see a JUnit test for that.

Please also make sure that the `hashCode()` method is creating a hash code which is usable in our case. The default method of the `Object` class for example is not helpful at all. Maybe you want to define an interface that takes a triple and returns a hash code so that we a) can replace the hash code method if we want to use a different method and b) it makes it much easier to test your method above because in the test you can use a dummy method that returns exactly the hash values that you need for your tests. It is hard to test things when relying on an unknown method for which you can not really predict the result.

Additionally: using hex values might be better (when you want to use Strings in the end instead of integers). To get rid of adding the additional zeros at the beginning of the string, you should check out the String.format method which can do stuff like that for you (and even can generate decimal, binary or hex representations of numbers)

Reply:
@MichaelRoeder  ´´Triple.hashcode()`` works like desired, it does not only compare the references. But I can still add an interface for computing a hash value for a single Triple.
Tests already exist in ´´deduplication.hashing.IntervalBasedMinHashFunctionTest``.
Regarding the bit shifting, we had some problems and therefore made this hacky solution :) but we will try again or come to you.

Reply:
I am not fully convinced that `Triple::hashCode()` is working as expected. It uses the following implementation:
```java
public static int hashCode( Node s, Node p, Node o ) 
        { return (s.hashCode() >> 1) ^ p.hashCode() ^ (o.hashCode() << 1); }
```
with `Node::hashCode()` defined as
```java
    @Override
    public int hashCode() 
        { return label.hashCode() * 31; }
```
However, the `label` can represent a URI, a literal or a blank node ID. In the latter case, you might have problems mapping two graphs to each other. Even if they are equal, their blank nodes might get different (random) IDs leading to different hash values. You may want to check the `hashCode()` function for this use case.

Reply:
Ignoring it in a first version is okay but shows the need for interfaces that allow the easy replacement of the hash algorithm since we already can foresee that the current approach has drawbacks that we would like to fix in the future :wink: